### DISCUSSION TITLE
Second derivatives of FIRST LAGRANGE

### DISCUSSION THREAD

**Discussion by mangerij** [idaholab/moose#21442] [Q&A General] [answered]
Hi, I want to be able to compute second derivatives of a variable in a postprocessor. This is an energy with a corresponding Kernel. For the Kernel we got around this by integrating by parts and putting one of the spatial derivatives on the test function. However, for accurate analysis of the energy we need the second derivatives of our variables. 

I am aware that second derivatives of FIRST LAGRANGE just return 0.0 because they are not stored but is there an easy way to track them?

edit: apologies if this was already asked in here, I tried search and couldn't find anything

**Comment by hugary1995** [comment, marked as answer]
The issue is that piecewise multilinear basis has zero second derivative. You'd want to project that variable onto a second order basis.

**Reply by mangerij** [reply]
Is the procedure to let an AuxVariable be SECOND MONOMIAL and then use an AuxKernel to computeValue to store the value of the FIRST LAGRANGE
If so, I suppose that worked but I find the derivatives to be _much much_ smaller than I expected.

**Reply by hugary1995** [reply]
Right, this won't work. By projecting something of first order onto a second order basis, you are essentially "adding information", and that additional information comes from nowhere. Think about the case where you want to make a line with constant slope (first order) become second order by adding an additional dof, that dof can be _anywhere_. For most conservative projections, you'll end up recovering the exact same straight line, hence the second derivative is very small.

Depending on your physics, you need to first figure out where that "additional information" may come from. Then there will be different ways of doing the projection. It's probably no longer appropriate calling it a projection, as projection typically loses information...

**Reply by GiudGiud** [reply]
You would need the variables that are used to compute the energy, likely the nonlinear variables, to be second order. Then you may compute a second order derivative. 

Otherwise you could build a stencil with more than one element to compute the second derivative. It's what we do in finite volume. It might not be super easy with the native coupleable (coupledValue etc) interface

**Reply by hugary1995** [reply]
@mangerij You can try something like this:
```python
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = 20
  second_order = true
[]

[Variables]
  [u]
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxVariables]
  [v]
    order = SECOND
    family = LAGRANGE
    [AuxKernel]
      type = NodalPatchRecoveryAux
      nodal_patch_recovery_uo = patch
      execute_on = 'TIMESTEP_END'
    []
  []
[]

[Kernels]
  [react]
    type = Reaction
    variable = u
  []
  [source]
    type = BodyForce
    variable = u
    function = 'sin(100*x)'
  []
[]

[Materials]
  [u]
    type = ParsedMaterial
    f_name = u
    args = u
    function = u
  []
[]

[UserObjects]
  [patch]
    type = NodalPatchRecoveryMaterialProperty
    patch_polynomial_order = SECOND
    property = u
    execute_on = 'TIMESTEP_END'
  []
[]

[Executioner]
  type = Transient
  num_steps = 1
[]

[Outputs]
  exodus = true
[]
```
This reconstructs a second order variable using a first order variable using the [Zienkiewicz-Zhu patch recovery algorithm](https://mooseframework.inl.gov/help/finite_element_concepts/nodal_patch_recovery.html)

**Reply by hugary1995** [reply]
The output looks like this:
![image](https://user-images.githubusercontent.com/31555204/176273474-16dc1715-6a85-4e8f-a1af-debd6bb2fb54.png)

**Comment by alimostafavi24** [comment]
I am also trying to do the same. Let's say I want to calculate the volume integral of d2u/dx2+d2u/dy2. Is it the right way to call it in the postprocessor kernel (indexing should be (0,0) and (1,1))?

`return _mu * (_u[_qp]) * (_second_u[_qp](0,0) + _second_u[_qp](1,1)  );`

Right now I am using a first order element for u. I might be able to use the second order element for my variable. If I do this, can the post processor return the integral correctly? It would be great if I can correctly integrate the second order derivative.

Thanks.

**Reply by GiudGiud** [reply]
If first order lagrange that will just be 0

are you using coupledSecond ? From the coupleable interface

**Comment by alimostafavi24** [comment]
All right. What is my best option here? Would using second order for variables be accurate or not? (here are the .h and .C files)

```
#pragma once

#include "ElementIntegralPostprocessor.h"
#include "MooseVariableInterface.h"

/**
 * This postprocessor computes a volume integral of the specified variable.
 *
 * Note that specializations of this integral are possible by deriving from this
 * class and overriding computeQpIntegral().
 */
class KEViscousWater : public ElementIntegralPostprocessor,
                                             public MooseVariableInterface<Real>
{
public:
  static InputParameters validParams();

  KEViscousWater(const InputParameters & parameters);

protected:
  virtual Real computeQpIntegral() override;

  /// Holds the solution at current quadrature points
  const VariableValue & _u;
  /// Holds the solution gradient at the current quadrature points
  const VariableGradient & _grad_u;

  const VariableValue & _vel_x;
  const VariableGradient & _grad_vel_x;
  const VariableSecond & _second_vel_x;

  const VariableValue & _c;
  const VariableGradient & _grad_c;
  
  const Real & _mu;
};
```

```
#include "KEViscousWater.h"

registerMooseObject("MooseApp", KEViscousWater);

InputParameters
KEViscousWater::validParams()
{
  InputParameters params = ElementIntegralPostprocessor::validParams();
  params.addRequiredCoupledVar("variable", "The name of the variable that this object operates on");
  params.addRequiredCoupledVar("vel_x", "vel_x");
  params.addRequiredCoupledVar("c", "c");

  params.addClassDescription("Computes a volume integral of the specified variable");
  params.addParam<Real>("mu", 1.0, "mu");
  return params;
}

KEViscousWater::KEViscousWater(
    const InputParameters & parameters)
  : ElementIntegralPostprocessor(parameters),
    MooseVariableInterface<Real>(this,
                                 false,
                                 "variable",
                                 Moose::VarKindType::VAR_ANY,
                                 Moose::VarFieldType::VAR_FIELD_STANDARD),
    _u(coupledValue("variable")),
    _grad_u(coupledGradient("variable")),
    _vel_x(coupledValue("vel_x")),
    _grad_vel_x(coupledGradient("vel_x")), 
    _second_vel_x(coupledSecond("vel_x")),
      

    _c(coupledValue("c")),
    _grad_c(coupledGradient("c")),
    _mu(getParam<Real>("mu"))


{
  addMooseVariableDependency(&mooseVariableField());
}

Real
KEViscousWater::computeQpIntegral()
{
  if (_c[_qp] < -0.0)
  return 0;
  return _mu * (_vel_x[_qp]) * (_second_vel_x[_qp](0,0) + _second_vel_x[_qp](1,1)  );
}

```

**Reply by GiudGiud** [reply]
integration by parts to reduce the order of the derivative is what we usually do

**Reply by alimostafavi24** [reply]
For me, I should do the volume integration.

How about creating a multiapp and use the LaplacianSplit? (I assume it gives d2u/dx2+d2u/dy2)

Thanks.

**Reply by GiudGiud** [reply]
I m not sure what you mean by that

Integration by parts is still a volume integration, just of a different term. And there's an additional boundary integral. But the reduction in order is worth it
